pkg/airgap/ociartifactsbundler.go

```diff
@@ -0,0 +1,452 @@
+/*
+Copyright 2024 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package airgap
+
+import (
+	"archive/tar"
+	"bytes"
+	"cmp"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"maps"
+	"net/http"
+	"os"
+	"path"
+	"slices"
+	"sync"
+
+	"github.com/containerd/platforms"
+	"github.com/k0sproject/k0s/internal/pkg/stringslice"
+	"github.com/k0sproject/k0s/pkg/k0scontext"
+
+	"github.com/distribution/reference"
+	"github.com/dustin/go-humanize"
+	"github.com/opencontainers/go-digest"
+	imagespecs "github.com/opencontainers/image-spec/specs-go"
+	imagespecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/registry"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+)
+
+type InsecureOCIRegistryKind uint8
+
+const (
+	NoInsecureOCIRegistry InsecureOCIRegistryKind = iota
+	SkipTLSVerifyOCIRegistry
+	PlainHTTPOCIRegistry
+)
+
+type RewriteRefFunc func(sourceRef reference.Named) (targetRef reference.Named)
+
+type OCIArtifactsBundler struct {
+	Log                   logrus.FieldLogger
+	InsecureRegistries    InsecureOCIRegistryKind
+	RegistriesConfigPaths []string // uses the standard Docker config if empty
+	PlatformMatcher       platforms.MatchComparer
+	RewriteTarget         RewriteRefFunc
+}
+
+func (b *OCIArtifactsBundler) Run(ctx context.Context, refs []reference.Named, out io.Writer) error {
+	var client *http.Client
+	if len := len(refs); len < 1 {
+		b.Log.Warn("No artifacts to bundle")
+	} else {
+		b.Log.Infof("About to bundle %d artifacts", len)
+		var close func()
+		client, close = newHttpClient(b.InsecureRegistries == SkipTLSVerifyOCIRegistry)
+		defer close()
+	}
+
+	creds, err := newOCICredentials(b.RegistriesConfigPaths)
+	if err != nil {
+		return err
+	}
+
+	newSource := func(ref reference.Named) oras.ReadOnlyTarget {
+		return &remote.Repository{
+			Client: &auth.Client{
+				Client:     client,
+				Credential: creds,
+			},
+			Reference: registry.Reference{
+				Registry:   reference.Domain(ref),
+				Repository: reference.Path(ref),
+			},
+			PlainHTTP: b.InsecureRegistries == PlainHTTPOCIRegistry,
+		}
+	}
+
+	copyOpts := oras.CopyOptions{
```

Comment:
IMO also worth extracting it to a separate function

Reply:
I think this is fine, as this is just a struct literal? Converting this into a function with a flat argument list would somehow feel wrong to me.

Reply:
That's fair enough